### PR TITLE
Update WordPress Monorepo

### PR DESCRIPTION
**Pull request by noahtallen**
#### Changes proposed in this Pull Request
Alternative to #50470 with our own changes, such as deduplication.

#### Testing instructions
Some items we need to verify which consume this on some level:
- [x] ETK
- [x] Gutenboarding
- [x] Blocks
- [x] Components
- [x] Calypso application
- [ ] others?

### CODE REVIEW

**Comment by sarayourfriend** [comment]
Smoke tested Gutenboarding, everything seems to work but someone from that team should probably take a closer look.

Also took a look through the WP components page in the devdocs and everything looks to be in good shape including our theme overrides.

Lots of unit test failures, however, some of which appear to be legitimate bugs in core: 

```
redirectToJetpack redirect needed 

TypeError: url.indexOf is not a function
    at removeQueryArgs (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@wordpress/url/build/@wordpress/url/src/remove-query-args.js:21:31)
    at getJetpackAuthorizeURL (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/client/my-sites/controller.js:616:14)
    at Object. (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/client/my-sites/checkout/checkout/test/maybe-jetpack-authorize.js:22:28)
    at Object.asyncJestTest (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise ()
    at mapper (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/queueRunner.js:75:41
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

This error appears many many times:

```
TypeError: MutationObserver is not a constructor
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/dom/dist/wait-for.js:92:18
    at new Promise ()
    at waitFor (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/dom/dist/wait-for.js:40:10)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/dom/dist/wait-for.js:174:54
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/react/dist/pure.js:58:22
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/react/dist/act-compat.js:60:24
    at batchedUpdates$1 (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/react-dom/cjs/react-dom.development.js:21856:12)
    at act (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/react-dom/cjs/react-dom-test-utils.development.js:929:14)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/react/dist/act-compat.js:59:20
    at new Promise ()
    at asyncAct (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/react/dist/act-compat.js:38:14)
    at Object.asyncWrapper (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/react/dist/pure.js:57:35)
    at waitForWrapper (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@testing-library/dom/dist/wait-for.js:174:35)
    at _callee$ (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/packages/shopping-cart/test/cart-actions.tsx:228:10)
    at tryCatch (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/regenerator-runtime/runtime.js:63:40)
    at Generator.invoke [as _invoke] (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/regenerator-runtime/runtime.js:293:22)
    at Generator.next (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/regenerator-runtime/runtime.js:118:21)
    at asyncGeneratorStep (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
    at _next (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
    at new Promise ()
    at Object. (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12)
    at Object.asyncJestTest (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise ()
    at mapper (/home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /home/teamcity-2/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-jasmine2/build/queueRunner.js:75:41
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

I'm not sure what to make of it, however 🤔

**Comment by sarayourfriend** [comment]
Okay I have a fix for the `indexOf` error. I think the test we wrote is actually bad, so I'm just updating the test.

**Comment by sarayourfriend** [comment]
Okay, @noahtallen and I spent some time 🍐 on the MutationObserver issue. It seems like the issue boils down to the fact that we've got two different `JSDom` environments being required into our project, one by `enzyme` and one by `testing-library`. Similar issues have been documented by other projects: https://github.com/testing-library/react-testing-library/issues/730

The suggested solution is to either: (a) include a polyfill for the MutationObserver API via the `mutation-observer` package in `calypso-build`'s jest setup; or (b) use the `resolutions` API for yarn to force JSDom to always resolve to the latest version.

Noah and I couldn't decide which of these was better so we're deferring to @scinos 😀 

Additional note: The reason Gutenberg hasn't run into this issue yet is because they don't employ the `waitFor...` methods from `testing-library` that employ the `MutationObserver`. Whatever we choose to do here, we might want to open a PR in core to replicate that to prevent anyone from running into this problem in the future. Of course, that means we should probably weight the polyfill option heavier because Gutenberg doesn't use `yarn` so the `resolutions` API is unavailable to them.

Anyway, what do you think @scinos?

**Comment by scinos** [comment]
Looks like adding an explicit dependency to `jest-environment-jsdom` fixed the issue.

As you said, this fails because we have two versions of `jsdom` in our dependency tree, imported by two different dependency chains:

* `jest-enzyme@7.1.2` -> ... -> `jest-environment-jsdom@24.9.0` -> `jsdom@11.12.0`
* `@automattic/calypso-build` -> `jest-config@26.6.3` -> ... -> `jsdom@16.5.0`

Yarn decides to hoist the former, so `node_modules/jsdom` points to `jsdom@11.12.0`.

However, [looking at Jest resolution for test environment](https://github.com/facebook/jest/blob/4ebbc8098760f56ae8a760a6ee8a4610c9a1b727/packages/jest-config/src/utils.ts#L170), it tries to resolve it relative to the project dir (in this case `./node_modules`) first, which results in it finding the wrong `jsdom`. If they would have started relative to jest itself (`./node_modules/jest-config/node_modules`) they would use the right one.

Adding it as an explicit dependency in `package.json` is a small hack to force `./node_modules/jsdom` to be right version.

File: .eslintrc.js [hunk 1]
```diff
@@ -53,20 +53,8 @@ module.exports = {
...
-					},
-				],
-				'no-restricted-modules': [
-					'error',
-					{
-						patterns: [ 'calypso/*' ],
-						message: "Packages shouldn't import from calypso",
-					},
-				],
+				'no-restricted-imports': [ 'error', { patterns: [ 'calypso/*' ] } ],
```
Comment (scinos) [inline .eslintrc.js:56]: [Doc](https://eslint.org/docs/rules/no-restricted-imports) says
> Please note that you may not specify custom error messages for restricted patterns as a particular module may match more than one pattern.

The new versions of eslint packages now enforce the config schema, it failed with the old config.

**Comment by scinos** [comment]
As this updates several eslint packages, I'm running a full eslint build to see if this would introduce new linting errors.

Edit: Yeah, it does. Two new errors:

image

**Comment by noahtallen** [comment]
~@scinos @sarayourfriend My apologies, I force-pushed after rebasing this on `trunk`, but forgot to pull the changes you pushed first. (So they are unfortunately missing now...) Could you add them again? Sorry about that!~

Edit: I believe we've added back everything now, but feel free to take another look to be sure.

**Comment by noahtallen** [comment]
Thanks for adding that again sara! I just pushed the eslint config fix, as will as a fix for the `target: _blank` problem.

File: client/layout/sidebar/item.jsx [hunk 2]
```diff
@@ -54,15 +54,16 @@ export default function SidebarItem( props ) {
...
+	const linkProps = showAsExternal ? { target: '_blank', rel: 'noreferrer' } : {};
+
 	return (
 		<li className={ classes } data-tip-target={ props.tipTarget } data-post-type={ props.postType }>
 			<a
 				className="sidebar__menu-link"
 				onClick={ props.onNavigate }
 				href={ url }
-				target={ showAsExternal ? '_blank' : null }
-				rel={ isExternalLink ? 'noopener noreferrer' : null }
```
Comment (noahtallen) [inline client/layout/sidebar/item.jsx:63]: Note: `noreferrer` includes `noopener` behavior according to the [HTML standard](https://html.spec.whatwg.org/multipage/links.html#link-type-noreferrer).

**Comment by noahtallen** [comment]
I've tried adding the dependency on jest-environment-jsdom by running `yarn add jest-environment-jsdom -W`. Let me know if that was the correct command, @scinos!

At this point, I believe this should be rebased, deduped, and ready for another look.

**Comment by scinos** [comment]
Rebased and deduped.

**Comment by sarayourfriend** [comment]
Can someone from @Automattic/luna take a look at this just to smoke test Gutenboarding once again?

**Comment by ciampo** [comment]
> Can someone from @Automattic/luna take a look at this just to smoke test Gutenboarding once again?

@sarayourfriend I just smoke tested Gutenboarding, lookin' good!


Are the changes in the bundle sizes expected?

**Comment by sarayourfriend** [comment]
> Are the changes in the bundle sizes expected?

I believe those are probably due to the new component system integration, though I didn't think they were actually being exported.

@gziolo Is the published `@wordpress/components` package being built with `COMPONENT_SYSTEM_PHASE` turned on since the latest releases?

**Comment by gziolo** [comment]
> @gziolo Is the published `@wordpress/components` package being built with `COMPONENT_SYSTEM_PHASE` turned on since the latest releases?

I think we assume by default it's disabled and in the Gutenberg plugin we enable it through `package.json`:
https://github.com/WordPress/gutenberg/blob/27cb4f6c7dd95cdb4b8aa03e43747c8d72763fc2/webpack.config.js#L126-L132

**Comment by fullofcaffeine** [comment]
How often do we get these from Renovate? Do we have to find a better tool if Renovate isn't working well enough? Thanks!

**Comment by fullofcaffeine** [comment]
The `phpunit` check for ETK is failing atm, but doesn't seem related to the changes. Here's the output:

```
mysqlcheck: Got error: 1045: Access denied for user 'username_here'@'172.18.0.5' (using password: YES) when trying to connect
2
✖ Error while running docker-compose command.
Creating efb2e59ed029579353b216b51132f603_cli_run ... 
Creating efb2e59ed029579353b216b51132f603_cli_run ... done
mysqlcheck: Got error: 1045: Access denied for user 'username_here'@'172.19.0.5' (using password: YES) when trying to connect
```

I have no idea why that's happening, though :(

**Comment by noahtallen** [comment]
Yeah, the ETK phpunit is failing because wp-env broke yesterday (more specifically, the upstream Docker image we rely on pushed a breaking change.). It has been solved in core (https://github.com/WordPress/gutenberg/pull/29800), but it won't get fixed for us here until it gets an npm update.

**Comment by sarayourfriend** [comment]
Confirmed that the size increases are due to G2 being present in `@wordpress/components`. Not sure how to fix this.

/cc @gziolo It doesn't seem like tree shaking is working for the current strategy... any ideas?

**Comment by gziolo** [comment]
> Confirmed that the size increases are due to G2 being present in `@wordpress/components`. Not sure how to fix this.
> 
> /cc @gziolo It doesn't seem like tree shaking is working for the current strategy... any ideas?

See my previous replay: https://github.com/Automattic/wp-calypso/pull/50922#issuecomment-797267830.

You need to update the webpack config add ensure `COMPONENT_SYSTEM_PHASE` is replaced with `0` so it can evaluate as `false`? Example from the WordPress core:

https://github.com/WordPress/wordpress-develop/blob/140bd894d6b7ef49575df761637db821e2fc27f0/tools/webpack/packages.js#L256-L257

**Comment by sarayourfriend** [comment]
I'm glad setting `COMPONENT_SYSTEM_PHASE` seems to have worked... but I'm still confused about this behavior.

I thought that the default for `COMPONENT_SYSTEM_PHASE` was 0? Should there be a bug opened if it's not defaulting to 0 correctly? It's always compared against `1` explicitly, why would that evaluate to `true` without it being set?

**Comment by noahtallen** [comment]
I bumped @wordpress/env back to v4. I think it had gotten changed back to v3 during a rebase, but it's needed to fix the phpunit tests :)

**Comment by sarayourfriend** [comment]
What's left on this PR @noahtallen? Do we just need to request reviews from some individual teams? Are those e2e test failures legit?

**Comment by noahtallen** [comment]
> What's left on this PR @noahtallen? Do we just need to request reviews from some individual teams? Are those e2e test failures legit?

Unsure about e2e tests, will investigate 👀 

Otherwise, i think just smoke testing everything which consumes these packages.

**Comment by noahtallen** [comment]
The e2e failure is a known broken test from what I can tell :)

**Comment by noahtallen** [comment]
I've tested ETK, gutenboarding (all the way through launch), and the calypso application and everything looks good so far. @sarayourfriend do you know what to test for blocks & components?

**Comment by sarayourfriend** [comment]
We should just smoke test the WP Components Gallery but honestly testing Gutenboarding is even better. I'll smoke test the gallery now.

Update: Components Gallery looks good to me.

**Comment by noahtallen** [comment]
I've run through the failing e2e's manually and confirmed they are still working. Possibly flaky, or not waiting long enough for the UI element to display

**Pull request merged**